pinot-core/src/main/java/org/apache/pinot/core/operator/combine/function/CombineFunction.java

```diff
@@ -0,0 +1,32 @@
+package org.apache.pinot.core.operator.combine.function;
+
+import org.apache.pinot.core.operator.blocks.results.BaseResultsBlock;
+
+
+public interface CombineFunction<T extends BaseResultsBlock> {
+
+  /**
+   * Merges a results block into the main mergeable results block.
+   * <p>NOTE: {@code blockToMerge} should contain the result for a segment without any exception. The errored segment
+   * result is already handled.
+   *
+   * @param mergedBlock The block that accumulates previous results. It should be modified to add the information of the
+   *                    other block.
+   * @param blockToMerge The new block that needs to be merged into the mergedBlock.
+   */
+  void mergeResultsBlocks(T mergedBlock, T blockToMerge);
+
+  /**
+   * Can be overridden for early termination. The input results block might not be mergeable.
+   */
+  default boolean isQuerySatisfied(T resultsBlock) {
+    return false;
+  }
+
+  /**
+   * Converts the given results block into a mergeable results block if necessary.
+   */
+  default T convertToMergeableBlock(T resultsBlock) {
```

Comment:
unclear if this is called once per `mergeResultsBlocks` or once at the very end. it would be nice to clarify that in the comment

Reply:
adjusted the javadoc. basically it is only called for the first argument to the `mergeResultsBlocks` function